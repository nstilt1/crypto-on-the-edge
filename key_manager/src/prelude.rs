pub use crate::{error::ProtocolError, Id};
#[cfg(feature = "server_client_ecdh_ecdsa_mode")]
pub use crate::{
    generated::server_client_ecdh_ecdsa_mode::*,
    server_client_ecdh_ecdsa_mode::key_manager::HttpPrivateKeyManager,
};

pub use private_key_generator::prelude::*;
pub use base64::alphabet::Alphabet;