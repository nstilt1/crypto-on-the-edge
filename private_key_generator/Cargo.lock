# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "autocfg"
version = "1.1.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d468802bab17cbc0cc575e9b053f41e72aa36bfa6b7f55e3529ffa43161b97fa"

[[package]]
name = "base16ct"
version = "0.2.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4c7f02d4ea65f2c1853089ffd8d2787bdbc63de2f0d29dedbcf8ccdfa0ccd4cf"

[[package]]
name = "base64ct"
version = "1.6.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8c3c1a368f70d6cf7302d78f8f7093da241fb8e8807c05cc9e51a125895a6d5b"

[[package]]
name = "blake2"
version = "0.11.0-pre.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3eb6b33ba68af672bcef0f6d1cceeeaf36e4143cd1456cafafda5d7f12d91f14"
dependencies = [
 "digest",
]

[[package]]
name = "block-buffer"
version = "0.11.0-pre.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3ded684142010808eb980d9974ef794da2bcf97d13396143b1515e9f0fb4a10e"
dependencies = [
 "crypto-common",
]

[[package]]
name = "cfg-if"
version = "1.0.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "baf1de4339761588bc0619e3cbc0120ee582ebb74b53b4efbf79117bd2da40fd"

[[package]]
name = "const-oid"
version = "0.10.0-pre.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f7e3352a27098ba6b09546e5f13b15165e6a88b5c2723afecb3ea9576b27e3ea"

[[package]]
name = "cpufeatures"
version = "0.2.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "53fe5e26ff1b7aef8bca9c6080520cfb8d9333c7568e1829cef191a9723e5504"
dependencies = [
 "libc",
]

[[package]]
name = "crypto-bigint"
version = "0.6.0-pre.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1943d7beadd9ce2b25f3bae73b9e9336fccc1edf38bdec1ed58d3aa183989e11"
dependencies = [
 "hybrid-array",
 "num-traits",
 "rand_core",
 "subtle",
 "zeroize",
]

[[package]]
name = "crypto-common"
version = "0.2.0-pre.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b7aa2ec04f5120b830272a481e8d9d8ba4dda140d2cda59b0f1110d5eb93c38e"
dependencies = [
 "getrandom",
 "hybrid-array",
 "rand_core",
]

[[package]]
name = "der"
version = "0.8.0-pre.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b489fd2221710c1dd46637d66b984161fb66134f81437a8489800306bcc2ecea"
dependencies = [
 "const-oid",
 "zeroize",
]

[[package]]
name = "digest"
version = "0.11.0-pre.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "065d93ead7c220b85d5b4be4795d8398eac4ff68b5ee63895de0a3c1fb6edf25"
dependencies = [
 "block-buffer",
 "const-oid",
 "crypto-common",
 "subtle",
]

[[package]]
name = "ecdsa"
version = "0.17.0-pre.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d7e045ee5c360512162782f3d4cb07d2f4ce8c4ef9bf7c77ec16d1cf60b3d5ca"
dependencies = [
 "digest",
 "elliptic-curve",
 "rfc6979",
 "signature",
]

[[package]]
name = "elliptic-curve"
version = "0.14.0-pre.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4a1775af172997a40c14854c3a9fde9e03e5772084b334b6a0bb18bf7f93ac16"
dependencies = [
 "base16ct",
 "base64ct",
 "crypto-bigint",
 "digest",
 "ff",
 "group",
 "hkdf",
 "hybrid-array",
 "pkcs8",
 "rand_core",
 "sec1",
 "serde_json",
 "serdect",
 "subtle",
 "zeroize",
]

[[package]]
name = "ff"
version = "0.13.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ded41244b729663b1e574f1b4fb731469f69f79c17667b5d776b16cda0479449"
dependencies = [
 "rand_core",
 "subtle",
]

[[package]]
name = "getrandom"
version = "0.2.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "190092ea657667030ac6a35e305e62fc4dd69fd98ac98631e5d3a2b1575a12b5"
dependencies = [
 "cfg-if",
 "libc",
 "wasi",
]

[[package]]
name = "group"
version = "0.13.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f0f9ef7462f7c099f518d754361858f86d8a07af53ba9af0fe635bbccb151a63"
dependencies = [
 "ff",
 "rand_core",
 "subtle",
]

[[package]]
name = "hkdf"
version = "0.13.0-pre.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "fd5d615ab5c462f96c309b3a00b19f373025a4981312f717f9df5bbd0201530c"
dependencies = [
 "hmac",
]

[[package]]
name = "hmac"
version = "0.13.0-pre.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ffd790a0795ee332ed3e8959e5b177beb70d7112eb7d345428ec17427897d5ce"
dependencies = [
 "digest",
]

[[package]]
name = "hybrid-array"
version = "0.2.0-rc.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "dcda354500b318c287a6b91c1cfbc42edd53d52d259a80783ceb5e3986fca2b2"
dependencies = [
 "typenum",
 "zeroize",
]

[[package]]
name = "itoa"
version = "1.0.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b1a46d1a171d865aa5f83f92695765caa047a9b4cbae2cbf37dbd613a793fd4c"

[[package]]
name = "libc"
version = "0.2.153"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9c198f91728a82281a64e1f4f9eeb25d82cb32a5de251c6bd1b5154d63a8e7bd"

[[package]]
name = "num-traits"
version = "0.2.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "da0df0e5185db44f69b44f26786fe401b6c293d1907744beaa7fa62b2e5a517a"
dependencies = [
 "autocfg",
]

[[package]]
name = "pkcs8"
version = "0.11.0-pre.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "935c09e0aecb0cb8f8907b57438b19a068cb74a25189b06724f061170b2465ff"
dependencies = [
 "der",
 "spki",
]

[[package]]
name = "ppv-lite86"
version = "0.2.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5b40af805b3121feab8a3c29f04d8ad262fa8e0561883e7653e024ae4479e6de"

[[package]]
name = "private_key_generator"
version = "0.1.0-pre.0"
dependencies = [
 "blake2",
 "ecdsa",
 "elliptic-curve",
 "hkdf",
 "rand_chacha",
 "rand_core",
 "sha2",
 "subtle",
 "zeroize",
]

[[package]]
name = "proc-macro2"
version = "1.0.78"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e2422ad645d89c99f8f3e6b88a9fdeca7fabeac836b1002371c4367c8f984aae"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.35"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "291ec9ab5efd934aaf503a6466c5d5251535d108ee747472c3977cc5acc868ef"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "rand_chacha"
version = "0.3.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e6c10a63a0fa32252be49d21e7709d4d4baf8d231c2dbce1eaa8141b9b127d88"
dependencies = [
 "ppv-lite86",
 "rand_core",
]

[[package]]
name = "rand_core"
version = "0.6.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ec0be4795e2f6a28069bec0b5ff3e2ac9bafc99e6a9a7dc3547996c5c816922c"
dependencies = [
 "getrandom",
]

[[package]]
name = "rfc6979"
version = "0.5.0-pre.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "045972f2f66b9467a2f6834b7fd0f9b23ca214b4a8700b880c36edb726e96da6"
dependencies = [
 "hmac",
 "subtle",
]

[[package]]
name = "ryu"
version = "1.0.17"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e86697c916019a8588c99b5fac3cead74ec0b4b819707a682fd4d23fa0ce1ba1"

[[package]]
name = "sec1"
version = "0.8.0-pre.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "02dc081ed777a3bab68583b52ffb8221677b6e90d483b320963a247e2c07f328"
dependencies = [
 "base16ct",
 "der",
 "hybrid-array",
 "pkcs8",
 "serdect",
 "subtle",
 "zeroize",
]

[[package]]
name = "serde"
version = "1.0.197"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3fb1c873e1b9b056a4dc4c0c198b24c3ffa059243875552b2bd0933b1aee4ce2"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.197"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "7eb0b34b42edc17f6b7cac84a52a1c5f0e1bb2227e997ca9011ea3dd34e8610b"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "serde_json"
version = "1.0.114"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c5f09b1bd632ef549eaa9f60a1f8de742bdbc698e6cee2095fc84dde5f549ae0"
dependencies = [
 "itoa",
 "ryu",
 "serde",
]

[[package]]
name = "serdect"
version = "0.3.0-pre.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "791ef964bfaba6be28a5c3f0c56836e17cb711ac009ca1074b9c735a3ebf240a"
dependencies = [
 "base16ct",
 "serde",
]

[[package]]
name = "sha2"
version = "0.11.0-pre.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8f33549bf3064b62478926aa89cbfc7c109aab66ae8f0d5d2ef839e482cc30d6"
dependencies = [
 "cfg-if",
 "cpufeatures",
 "digest",
]

[[package]]
name = "signature"
version = "2.3.0-pre.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1700c22ba9ce32c7b0a1495068a906c3552e7db386af7cf865162e0dea498523"
dependencies = [
 "digest",
 "rand_core",
]

[[package]]
name = "spki"
version = "0.8.0-pre.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cb2b56670f5ef52934c97efad30bf42585de0c33ec3e2a886e38b80d2db67243"
dependencies = [
 "base64ct",
 "der",
]

[[package]]
name = "subtle"
version = "2.5.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "81cdd64d312baedb58e21336b31bc043b77e01cc99033ce76ef539f78e965ebc"

[[package]]
name = "syn"
version = "2.0.52"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "b699d15b36d1f02c3e7c69f8ffef53de37aefae075d8488d4ba1a7788d574a07"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "typenum"
version = "1.17.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "42ff0bf0c66b8238c6f3b578df37d0b7848e55df8577b3f74f92a69acceeb825"

[[package]]
name = "unicode-ident"
version = "1.0.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3354b9ac3fae1ff6755cb6db53683adb661634f67557942dea4facebec0fee4b"

[[package]]
name = "wasi"
version = "0.11.0+wasi-snapshot-preview1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9c8d87e72b64a3b4db28d11ce29237c246188f4f51057d65a7eab63b7987e423"

[[package]]
name = "zeroize"
version = "1.7.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "525b4ec142c6b68a2d10f01f7bbf6755599ca3f81ea53b8431b7dd348f5fdb2d"
