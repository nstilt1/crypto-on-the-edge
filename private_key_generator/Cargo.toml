[package]
name = "private_key_generator"
version = "0.1.0-pre.0"
edition = "2021"
rust-version = "1.73"
description = "Private key generator, based on HKDF"
readme = "README.md"
keywords = [
    "crypto",
    "KDF"
]
categories = [
    "cryptography",
    "no-std"
]
license = "MIT OR Apache-2.0"

[lib]
doctest = false # need to change this before publishing

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
ecdsa = { version = "0.17.0-pre.5", features = ["signing"]}
elliptic-curve = { version = "0.14.0-pre.5", features = ["ecdh", "jwk"] }
hkdf = "=0.13.0-pre.3"
init_boxed = { path = "../init_boxed" }
rand_chacha = { version = "0.3" }
rand_core = "0.6"
subtle = "2"
zeroize = "1"

[dev-dependencies]
blake2 = "0.11.0-pre.3"
#p256 = { version = "0.14.0-pre.0" }
rand = "0.8"
sha2 = "0.11.0-pre.3"
trybuild = "1"

[features]
alloc = ["init_boxed/alloc"]
default = ["std"]
std = ["hkdf/std"] # std is required for checking timestamps against UNIX_EPOCH