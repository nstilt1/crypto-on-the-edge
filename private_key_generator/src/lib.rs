//! This library handles the generation of private keys, has capabilities for
//! expiring private keys, and eliminates the need to store private keys.
//!
//! <div class="warning">
//! Security notice: This is not a replacement for a Hardware Security Module,
//! and this crate has not been independently audited or reviewed. </div>
//!
//! [KeyGenerator](crate::KeyGenerator) is essentially a wrapper around an
//! HMAC-based Key Derivation function ([HKDF](hkdf::Hkdf)) with some utility
//! functions, along with a [BinaryId](crate::BinaryId) struct. The ID contains
//! an HMAC for determining probable[^note] authenticity of a user-supplied ID.
//! The ID could be an identifier for an ECDH key, an ECDSA key, or a client's
//! ID. The ID can also be encoded with an expiration timestamp and a version.
//!
//! [^note] [BinaryId](crate::BinaryId) contains an `HmacLen` parameter, which
//! determines the ID's HMAC's bits of security when validating IDs.
//!
//! Falsely validated IDs should not be critical to the security of your
//! application. The validation is primarily for preventing unnecessary database
//! read operations or other processing just to find out that a supplied ID was
//! invalid. If a "client ID" was falsely validated, you probably won't find the
//! ID in your database. If an ECDH Key ID was falsely validated, you will not
//! be able to decrypt the message, or they will not be able to decrypt the
//! message you send to them. If an ECDSA Key was falsely validated, then an
//! arbitrary signature might be created, and the attacker should not be able to
//! obtain any information from it.
//!
//! # Example
//! Before we can begin to use our generated private keys, we need to distribute
//! the public keys and key IDs. You can distribute the keys several different
//! ways:
//! 1. If you aren't concerned about timing attacks, you could create a REST API
//!    method that returns a public key and a key ID.
//! 2. You could also create a static webpage that you periodically update with
//!    your latest public keys and key IDs.
//! 3. If you are only using these keys with your own network of machines, you
//!    could theoretically use the same HMAC key and application ID to generate
//!    compatible private and public keys without needing to distribute or store
//!    public keys.
//! ```
//! use p256::NistP256;
//! use sha2::Sha256;
//! use private_key_generator::{
//!     Id, typenum::consts::{U48, U4}, KeyGenerator
//! };
//! use zeroize::Zeroize;
//!
//! // an EPOCH reference point for encoding timestamps with only 3 bytes
//! const EPOCH: u64 = 1709349508;
//! type EccKeyIdV1 = Id<
//!     U48,   // binary encoded ID length
//!     U4,    // binary encoded HMAC length
//!     6,     // max prefix length; use multiple of 3 for Base64 representation
//!     0,     // Info byte offset
//!     1,     // version number
//!     5,     // number of bits used to represent the version number
//!     EPOCH
//! >;
//!
//! /// You can make a private function for initializing a key generator, and
//! /// you could use the `prk` output with `from_prk` instead of `new` or `extract` to save a small amount of time.
//! fn initialize_key_generator() -> KeyGenerator {
//!    KeyGenerator::<Sha256>::new(&[42u8; 64], b"arbitrary application ID");
//! }
//!
//! // a function that generates some P-384 public keys and IDs with an expiration timestamp. You will need to distribute them as suggested above
//! fn generate_pubkeys() {
//!     let key_generator = initialize_key_generator();
//!     let expiration_date = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() + 60 * 60 * 24 * 365;
//!     let (id, pubkey) = key_generator.generate_ecdsa_key_and_id::<NistP256, EccKeyIdV1>(&[], Some(expiration_date), None, &mut OsRng);
//!     let (ecdh_key_id, pubkey) = key_generator.generate_ecdh_pubkey_and_id::<NistP256, EccKeyIdV1>(&[], Some(expiration_date), None, &mut OsRng);
//! }
//!
//! // a function that makes or uses private keys generated from IDs
//! fn use_key_ids() {
//!     let key_generator = initialize_key_generator();
//!     let client_id: &[u8];
//!     let client_provided_ecdh_key_id: &[u8];
//!     let client_pubkey: &[u8];
//!     let encrypted_data: &[u8];
//!     if let Ok(id) = key_generator.validate_ecdh_key_id::<EccKeyIdV1>(client_provided_ecdh_key_id, None) {
//!         let pubkey = PublicKey::<NistP384>::from_sec1_bytes(client_pubkey).unwrap();
//!         let shared_secret = key_generator.ecdh_using_key_id::<NistP256>(id, None, pubkey);
//!     }
//!     
//!     let client_provided_ecdsa_key_id: &[u8];
//!     if let Ok(id) = key_generator.validate_ecdsa_key_id::<NistP256>(client_provided_ecdsa_key_id, None) {
//!         let private_key = key_generator.generate_ecdsa_key_from_id::<NistP256>(id, None);
//!     }
//! }
//! ```
//! Stuff for the ID
//! has a generic argument that determines the length of the `HMAC` within the
//! ID itself. The longer the HMAC, the less likely a forged ID will make it
//! through. The `HMAC` does not need to be very long. If a forged Client ID
//! made it through this validation, then you shouldn't be able to find it in
//! your database. If a forged ECDH Key ID made it through, then shouldn't be
//! able to decrypt their message. If a forged ECDSA Key ID made it through,
//! then the attacker will receive a signature from an arbitrary private ECDSA
//! key, and since signatures are designed to be public, this should not reveal
//! any information about the private key generation.

#![no_std]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![forbid(unsafe_code)]
#![warn(missing_docs, rust_2018_idioms)]

#[cfg(feature = "std")]
extern crate std;

pub use ecdsa;
pub use hkdf::hmac::digest::typenum;
pub use hkdf::hmac::digest::Digest;

pub use elliptic_curve;
pub use hkdf;

mod id;
mod key_generator;

mod error;
mod traits;
mod utils;

pub use error::InvalidId;
pub use id::BinaryId;
pub use key_generator::{KeyGenerator, SimpleKeyGenerator};
pub use traits::{CryptoKeyGenerator, EncodedId};
