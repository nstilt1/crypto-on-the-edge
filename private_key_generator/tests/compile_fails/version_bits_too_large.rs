use private_key_generator::prelude::*;
use hkdf::hmac::Hmac;
use rand_chacha::ChaCha8Rng;
use rand::rngs::OsRng;
use sha2::Sha256;

type TestId = BinaryId<
    U48, // IdLength: okay
    U5,  // MacLength: okay
    5,   // MAX_PREFIX_LEN: okay
    use_timestamps::Sometimes
>;

type InvalidVersionBitsConfig = VersioningConfig<
    0,              // EPOCH
    1_000_000_000,  // VERSION_LIFETIME
    33,             // VERSION_BITS
    56,             // TIMESTAMP_BITS
    4,              // TIMESTAMP_PRECISION_LOSS is too high because TIMESTAMP_BITS + TIMESTAMP_PRECISION_LOSS is above 64
    1_000_000_000,  // MAX_KEY_EXPIRATION_TIME
>;

fn main() {
    type K = KeyGenerator<Hmac<Sha256>, InvalidVersionBitsConfig, ChaCha8Rng, Sha256>;

    let mut k = K::new(&[48u8; 32], b"ff");
    let _id = k.generate_keyless_id::<TestId>(&[], &[], None, None, &mut OsRng);
}