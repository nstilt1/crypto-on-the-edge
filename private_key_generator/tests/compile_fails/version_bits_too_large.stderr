error[E0080]: evaluation of `<private_key_generator::VersioningConfig<0, 1000000000, 33, 56, 4, 1000000000, 800> as private_key_generator::VersionConfig>::VERSION_BITS` failed
 --> src/key_generator.rs
  |
  |                     [/* VERSION_BITS must be less than or equal to 32 */][VERSION_BITS as usize]
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ index out of bounds: the length is 0 but the index is 33

note: erroneous constant used
   --> src/id.rs
    |
    |                 &[V::VERSION_BITS, V::TIMESTAMP_BITS],
    |                   ^^^^^^^^^^^^^^^

note: erroneous constant used
   --> src/id.rs
    |
    |                 &[V::VERSION_BITS],
    |                   ^^^^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn <private_key_generator::BinaryId<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UTerm, private_key_generator::prelude::B1>, private_key_generator::prelude::B1>, private_key_generator::prelude::B0>, private_key_generator::prelude::B0>, private_key_generator::prelude::B0>, private_key_generator::prelude::B0>, private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UTerm, private_key_generator::prelude::B1>, private_key_generator::prelude::B0>, private_key_generator::prelude::B1>, 5, private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UInt<private_key_generator::typenum::UTerm, private_key_generator::prelude::B1>, private_key_generator::prelude::B1>, private_key_generator::prelude::B1>, private_key_generator::prelude::B0>, private_key_generator::prelude::B1>, private_key_generator::prelude::B0>> as private_key_generator::EncodedId>::generate::<private_key_generator::VersioningConfig<0, 1000000000, 33, 56, 4, 1000000000, 800>>`
   --> src/key_generator.rs
    |
    |           let (mut id, trimmed_timestamp) = Id::generate::<V>(
    |  ___________________________________________^
    | |             prefix,
    | |             expiration,
    | |             associated_data.is_some(),
    | |             self.current_version_epoch,
    | |             rng,
    | |         )?;
    | |_________^

note: erroneous constant used
   --> src/key_generator.rs
    |
    |                 &[V::VERSION_BITS, V::TIMESTAMP_BITS],
    |                   ^^^^^^^^^^^^^^^

note: erroneous constant used
   --> src/key_generator.rs
    |
    |                 &[V::VERSION_BITS],
    |                   ^^^^^^^^^^^^^^^
